src/xds.rs

```diff
@@ -171,6 +172,17 @@ impl ProxyStateUpdateMutator {
                     .remove_endpoint(&prev.uid, &endpoint_uid(&prev.uid, None));
             }
 
+            // if expected is true we are removing a workload, we should forget it
+            // if expected is false this is an update, no need to forget on update
+            if expected
```

Comment:
maybe_remove is called on insert_workload so this seems problematic

Reply:
Also we happen to call removes after adds (see HandlerWrapper), but is it guaranteed? Maybe we can add tests to ensure if we get a XDS update with [add SA A, remove SA A] we do not end up flipping the cert

Reply:
mabye_remove sets expected to `false`. The check for expected stops forgetting when during insert, yeah?

Agree on tests. I started adding some but wound up at a dead end so decided to PR some code before spending more time on it incase this path received a firm "no"